packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -0,0 +1,125 @@
+} from './folding-state';
+
+interface ResultWithFolding extends Result {
+  parentResult?: ResultWithFolding;
+  childResults?: ResultWithFolding[];
+}
+
+function getRawValue(result: ResultWithFolding, field: string) {
+  const rawValue = result.raw[field] as string | [string] | undefined;
+  if (isArray(rawValue)) {
+    return rawValue[0];
+  }
+  return rawValue;
+}
+
+function areDefinedAndEqual<T>(
+  value1: T | undefined,
+  value2: T | undefined
+): boolean {
+  return (value1 || value2) !== undefined && value1 === value2;
+}
+
+function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
+  const results = [result];
+  if (result.parentResult) {
+    results.unshift(result.parentResult);
+  }
+  if (result.childResults?.length) {
+    results.push(
+      ...result.childResults.flatMap((childResult) =>
+        flattenResultTree(childResult)
+      )
+    );
+  }
+  return results;
+}
+
+function foldResult(
+  source: ResultWithFolding,
+  results: ResultWithFolding[],
+  fields: FoldingFields
+): FoldedResult {
+  const sourceChildValue = getRawValue(source, fields.child);
+  return {
+    ...source,
+    children: sourceChildValue
+      ? results
+          .filter(
+            (result) =>
+              getRawValue(result, fields.child) !==
+                getRawValue(source, fields.child) &&
+              getRawValue(result, fields.parent) === sourceChildValue
+          )
+          .map((result) => foldResult(result, results, fields))
+      : [],
+  };
+}
+
+function createCollection(
+  relevantResult: ResultWithFolding,
+  fields: FoldingFields
+): FoldedResult {
+  const flattenedResults = removeDuplicates(
+    flattenResultTree(relevantResult),
+    (result) => result.uniqueId
+  );
+
+  const topLevelResult =
+    flattenedResults.find(
+      (result) =>
+        getRawValue(result, fields.parent) === undefined ||
+        areDefinedAndEqual(
+          getRawValue(result, fields.parent),
+          getRawValue(result, fields.child)
+        )
+    ) ?? relevantResult;
+
+  return foldResult(topLevelResult, flattenedResults, fields);
+}
+
+function createCollections(
+  results: ResultWithFolding[],
+  fields: FoldingFields
+): FoldedResult[] {
+  return results.map((result) => createCollection(result, fields));
+}
+
+export const foldingReducer = createReducer(
+  getFoldingInitialState(),
+  (builder) =>
+    builder
+      .addCase(executeSearch.fulfilled, (state, {payload}) => {
+        state.collections = state.enabled
+          ? createCollections(payload.response.results, state.fields)
+          : [];
+      })
+      .addCase(fetchMoreResults.fulfilled, (state, {payload}) => {
+        state.collections = state.enabled
+          ? [
+              ...state.collections,
+              ...createCollections(payload.response.results, state.fields),
+            ]
+          : [];
+      })
+      .addCase(registerFolding, (state, {payload}) => ({
+        enabled: true,
```

Comment:
Similar to other registration actions, we should add an `if` check to only update the state if `enabled` is `false`. This would prevent a person instantiating multiple folding controllers with different fields  from overriding an breaking the initial instances.